### ISSUE TITLE
Streaming JSON unmarshalling of any nested structure

### ISSUE THREAD

**Issue by markusthoemmes** [akka/alpakka#841]
When consuming public APIs and trying to embrace streaming I constantly face issues when it comes to unmarshalling JSON. Many many APIs (CouchDB for instance) do not provide me with a "clean" stream of data in form of a JSON array or a JSON object. They rather nest the data that I would like to get the most in some structure of metadata, which could look like this:

```json
{
  "count": 10,
  "rows": [...]
}
```

(This is of course a very simple case, other APIs are much worse, but this isn't about API bashing).

In the above case, https://github.com/knutwalker/akka-stream-json won't help me to parse `rows` in a streaming fashion because jawn only supports a stream of json-objects, a single array or a single value when it comes to the async parser.

I stumbled upon https://github.com/jsurfer/JsonSurfer which implements **exactly** the functionality I'd need to extract (streamingly) exactly the data that I need. I already built a PoC of a `GraphStage` which will support the following interface:

```scala
Flow[ByteString, JValue, NotUsed].via(JsonStream.flow[JValue]("$.rows[*]")
```

The above would now stream every element of that array as it occurs, and drop anything else immediately. JsonSurfer supports even more sophisticated models of filtering the stream even narrower, which is icing on the cake.

If this is deemed useful (it has some overlap with #320, but that doesn't seem to hit the ground anytime soon sadly) I'll straight away open a pull-request and work on productizing this piece of work. I myself cannot count how many times I wished this existed OOTB 😋

**Comment by ennru** [comment]
Would it make sense to select the data array and than stream its content as `ByteString` ignoring the rest of the Json structure?
That way JsonFraming and any Json parser can be applied to that data.

**Comment by markusthoemmes** [comment]
@ennru In theory, sure. In practice though, `JsonSurfer` already parses the JSON, that'd result in duplicate work.

I've written a generic version of this though, you can plugin whatever AST you want to. In reality the syntax is:

```scala
Flow[ByteString, J, NotUsed].via(JsonStream.flow[J]("$.rows[*]")
```

Where there is an implementation of `JsonSupport` for `J` like so:

```scala
trait JsonSupport[T] {
  implicit val support: JsonSupport[T]
  val provider: JsonProvider[_, _, _]
}
```

`JsonProvider[_, _, _]` is a native Jackson interface, so that'll be used under the hood.

Regarding where to put the code: This will have a dependency on both `JsonSurfer` (MIT licensed) and `Jackson` (Apache 2.0 licensed). Would it make sense to have the code in this repository as a subproject or would it make more sense to open a dedicated repository and link it in as an external library?

**Comment by ennru** [comment]
Ok, sounds good.
I'd like to see a Json module in Alpakka and using JsonSurfer seems to fit our needs. As you said #320 is not moving and event-based parsing of Json would be very interesting as well.
(I've a few experiments of streaming XML to Json conversion lurking in a branch.)

So, please: give it a go!

**Comment by kroeberk** [comment]
Hi @markusthoemmes,

Is it possible to see your proof of concept code somewhere? I'm working on a project with a ton of huge JSON responses and I desperately need akka + json event based processing. I would like to experiment with your code. Any chance you make it public? It would be greatly appreciated.

~Karl

**Comment by johanandren** [comment]
I'm not sure what happened with #63 but AFAIR that also contained streaming JSON parsing/extracting without collecting the whole document in memory.

**Comment by kroeberk** [comment]
@markusthoemmes Thank you Markus. It doesn't have to be a proper pull request, I would also take a text dump or anything. It's so annoying that there's not proper event based json streamer, like with xml.

**Comment by markusthoemmes** [comment]
@kroeberk I finally got around to open a PR. It's rough, but it should work. I opted for the version that just pushes ByteStrings downstream, so I don't have to fiddle with depending on JSON libraries and so on.

See: https://github.com/akka/alpakka/pull/899

**Comment by ennru** [comment]
This is now implemented in #899 and available in Alpakka 0.19.
